### PR TITLE
Update Container Wait

### LINKED ISSUE #32242: `docker run` can't rely on the Events API.
**Description**

Since Docker 1.13, the implementation of `docker run ...` changed to use the `Events` API over `ContainerWait` to wait for the container to exit. This was done to facilitate the move of handling `autoRemove` on the daemon rather than the CLI - the problem with the Wait API being that it only waited for the container to exit, and not for the container to be removed which would cause race conditions if, for example, a successive CLI command tried to create a container with the same name and it had not yet been cleaned up by the daemon.

While there is no problem with this Events API approach when connecting directly to the docker daemon 
which is running the container, I have seen several issue with it when the "daemon" is actually a Docker Swarm cluster (classic swarm, not swarmkit). One major problem is that swarm does not support event filtering at all, and so it ignores the container filter and ends up streaming back all cluster events. This means that your CLI may receive a "die" event for another container, causing your CLI to exit pre-maturely. Even if filtering were added to the swarm manager, there is yet another issue where its aggregated event stream is not reliable - it may temporarily lose the event stream from the engine which is running the target container but will not disconnect the client because the aggregate stream as a whole is still working. If the container exits during this short disconnection window, the events will never be sent to the CLI and it will hang forever. There is no way to guarantee that the event will be sent back to the client.

### PR DESCRIPTION
**Pull request by jlhawn**
**Summary**

The goal for this PR is to be able to use the `POST /containers/{name}/wait` API in order to no longer rely on the Events API when doing `docker run ...` and `docker start` in attached mode. Here is the set of behavioral changes in the API and CLI:

- The wait API now has on optional query parameter, `condition`, which specifies the state of the container which the client would like to wait for:

    - **`not-running`**

        This is the default value which maintains the existing behavior of the Wait API. It blocks if the container is in a state which is considered to be *running*, i.e., any of *running*, *paused*, or *restarting*. If it is already not in any of those states, then it would not block at all and return immediately with the current exit code.

    - **`next-exit`**

        This condition allows the client to wait for the container's state to *change* to a not-running state or to be removed. This is different from the *not-running* option because it will block if the container is in the *created*, *exited*, *dead*, or *removing* state, blocking through a transition to one of the running states and back to one of *exited* states or until the container is removed, whichever comes first. It is similar to waiting for a *die* or *destroy* event. This option is very useful to use before issuing a container start API request.

    - **`removed`**

        This condition allows the client to wait for the container to be removed. It is similar to waiting for a *destroy* event.

- The Wait API will now return HTTP response headers immediately. This serves an an acknowledgement that a wait has been registered. The JSON response body will not be returned until the wait condition is met.

- The Wait API's request context is extended to wait for a `http.CloseNotifier` event in the case that the client closes the connection while the server is blocked on the wait condition. This prevents leaked goroutines but should only be necessary until we switch to using Go 1.8's request context which cancels automatically if the client disconnects.

- ~~To handle a manual detach (when the user enters the detach keys sequence when attached in TTY mode), the CLI now detects this key sequence, stops sending stdin, and causes the CLI to exit immediately. This is done as an alternative to waiting for a *detach* event.~~ The rug has been pulled out from under this bullet point. This will have to be done in the github.com/docker/cli repo along with the other CLI changes.

Follow-up to #31794

Fixes #32242

### CODE REVIEW

**Comment by tonistiigi** [comment]
@WeiZhang555 @cpuguy83 any concerns for the approach?

**Comment by WeiZhang555** [comment]
The design looks good, I like it :-) 
https://github.com/docker/docker/pull/31744 seems need to reimplement based on this, it could be simplified a lot with this PR.

**Comment by cpuguy83** [comment]
@WeiZhang555 Indeed, however we still need to support old daemons.

**Comment by WeiZhang555** [comment]
@cpuguy83 Yep, you're right, then we have to keep both `waitExitOrRemoved` and this, backward compatibility makes this so complicated :cry:

**Comment by jlhawn** [comment]
Well, at least `janky` seems happy with this first commit. I'll have to look into the failures with `windowsRS1` and `z` tests later. I'll push the second commit now.

**Comment by jlhawn** [comment]
I've removed the *[WIP]* label from this PR and make the necessary changes to the API client and docs. Please take a look. I've also the `waitExitOrRemoved()` function in a way that maintains compatibility with older versions.

While all of the integration tests pass locally for me, I have seen some flakiness in a few tests that Jenkins runs. They mostly get this error:

```
Cannot kill container {ContainerID}: Container {ContinerID} is not running
```

But I can't tell if it's related to my changes or not since I tried not to modify container state/lifecycle in any way.

File: cli/command/container/utils.go
Comment (jlhawn) [inline cli/command/container/utils.go]: Oops, I forgot to send do `statusC <- 125` here.

**Comment by jlhawn** [comment]
I've hit an issue on something that I forgot to consider: `docker run -it ...` where the user enters the `DetachKeys`. In this case, the `ContainerAttach` API call returns but we still block on waiting for the container to exit, which it never will. This is why the `DockerSuite.TestAttachAfterDetach` integration test is failing everywhere.

The current implementation of `waitExitOrRemoved()` handles this by waiting for a `detach` event to unblock and return a `0` exit code. That is also not perfect because you have no way of knowing which CLI session the detach came from. I also want to completely avoid using the events API. I could overload the wait API with another option to wait for a detach, but you still don't know which attach request it was associated with.

One solution that I can think of would be to do an inspect of the container after the attach returns. If the container is still running then we can assume that the attach connection ended because the user entered the detach control keys. If the container has exited or is not found then we can assume that the attach connection ended because the container exited and/or was removed. The only problem with this is that there is a very small race window after the container exits between when the container's IO streams are closed (which causes the attach request to end) and the container's state is set to stopped/restarting. If the inspect request comes in after the attach connection closes but somehow before the container's state is marked as stopped then it would be interpreted as a manual detach since it would look like the container is still running.

Another clever solution to detecting the detach control keys is to actually do that detection client-side, passing the stream through the logic that tries to detect the detach control keys in the same way that the daemon-side does. Then we'll know for sure that the client initiated the detach.

**Comment by jlhawn** [comment]
Windows tests look good. There's just a race condition in one of the DockerSuite integration tests. I've opened an issue for it. #32306

**Comment by jlhawn** [comment]
The powerpc also looks good except for the same flaky test. It also failed on another flaky test: `TestSwarmIngressNetwork`.

**Comment by jlhawn** [comment]
The `z` tests had the same single flaky test failure as on Windows: `TestRmRestartingContainer`.

**Comment by jlhawn** [comment]
I've decided to carry a fix for `TestRmRestartingContainer` on top of this PR to try to get all of the tests to pass. I've also got another PR open with just that commit: https://github.com/docker/docker/pull/32307

**Comment by jlhawn** [comment]
Aw Yiss. ✅✅✅✅✅✅

**Comment by jlhawn** [comment]
I've edited the initial PR comment above to summarize all of the changes.

**Comment by jlhawn** [comment]
I've rebased after a conflict with `docs/api/version-history.md`.

**Comment by jlhawn** [comment]
Since this didn't make the cut for 17.05 and API 1.29 I've had to add on an extra commit which bumps the version checks I added to use 1.30 instead and update `docs/api/version-history.md`.

File: cli/command/container/wait.go
Comment (tonistiigi) [inline cli/command/container/wait.go]: Is there an use case for cli options? We don't necessarily need to expose the whole api in cli interface.

**Comment by tonistiigi** [comment]
Design LGTM

ping @WeiZhang555 for review. Tentatively moving to code-review

File: cli/command/container/wait.go
Comment (jlhawn) [inline cli/command/container/wait.go]: I remember doing something in the past where I wanted to wait for a container on the CLI that was only "created" but I can't seem to remember why.

If we think it's just clutter then I can remove it, it would also remove the man-page changes and  some integration-cli tests that I added.

Comment (WeiZhang555) [inline cli/command/container/wait.go]: > **not-running**
> It blocks if the container is in a state which is considered to be running, i.e., any of running, paused, or restarting.

From your statement, I would expect that `docker restart` a running container won't make `wait --condition not-running` return, but actually it's not. After reading your codes, I see that `not-running` shares same `waitStop` channel with `next-exit`, the only difference is handling of initial state of container.

I think `not-running` is enough, not sure if there's any scenario for `next-exit`.

Comment (jlhawn) [inline cli/command/container/wait.go]: `docker restart ...` and restart policies technically causes an atomic `stop` followed by a `start` (with a timeout before actually starting the process in the case of restart policies).

File: cli/command/container/hijack.go
Comment (cpuguy83) [inline cli/command/container/hijack.go]: We need to capture this default somewhere.

File: client/container_wait.go
Comment (cpuguy83) [inline client/container_wait.go]: waits waits

File: pkg/term/proxy.go
Comment (cpuguy83) [inline pkg/term/proxy.go]: Probably not the best naming here.
`NewEscapeProxy`? `WithEscape`?

File: pkg/term/proxy.go
Comment (cpuguy83) [inline pkg/term/proxy.go]: `escapeProxy`?

File: container/state.go
Comment (cpuguy83) [inline container/state.go]: 🤦‍♂️

File: container/state.go
Comment (cpuguy83) [inline container/state.go]: Maybe just need exited and removed?

**Review by cpuguy83 — CHANGES_REQUESTED** [review]
Seems mostly alright.

File: cli/command/container/hijack.go
Comment (cpuguy83) [inline cli/command/container/hijack.go]: e.g., maybe package level var.

File: api/server/router/container/container_routes.go
Comment (jlhawn) [inline api/server/router/container/container_routes.go]: Thanks, this is good to know.

File: container/state.go
Comment (jlhawn) [inline container/state.go]: What is the definition of "exited" though?.

The original implementation of `wait` considered a container which was in the `"created"` state to be *exited* (`wait` for a `"created"` container returns with exit status 0 immediately) but we explicitly want to block when the container has not yet started, both in the case of `docker run ...` and `docker start -a ...` where the container may either be created or even already exited and you are about to manually start it again. The `next-exit` condition is used in these two cases. You can think about the `next-exit` condition as a state transition, or as the equivalent of waiting for a `die` event for the container.

File: container/state.go
Comment (jlhawn) [inline container/state.go]: I don't think this method is even used anymore, so I can just remove it.

**Comment by jlhawn** [comment]
@cpuguy83 I've rebased and adde 3 commits to address your comments, please take a look.

**Comment by jlhawn** [comment]
@tonistiigi @WeiZhang555 I've also added a commit which removes the changes to the CLI, so there's no longer a `--condition` flag and it just keeps the old behavior (i.e., `--condition=not-running`). I also removed the integration tests which covered those CLI options.

**Comment by jlhawn** [comment]
I used to be able to change labels on the `docker/docker` repo but now I can't. Can someone trigger a rebuild on `powerpc`, please?

File: client/container_wait.go
Comment (cpuguy83) [inline client/container_wait.go]: It would be nice to consolidate this with the new function.
The new implementation looks more correct assuming we get a response immediately in both cases.

File: cli/command/container/utils.go
Comment (cpuguy83) [inline cli/command/container/utils.go]: Can we define these conditions in api/types?

**Review by cpuguy83 — CHANGES_REQUESTED** [review]

File: client/container_wait.go
Comment (jlhawn) [inline client/container_wait.go]: We can't use the new client implementation because it makes the request and blocks on response headers *before* spawning a goroutine to read the body. There is also logic on the server side which will not return headers immediately if the API version is before `1.30`.

Even if the new server implementation did always return headers immediately, regardless of the API version, this would cause problems when talking to older servers which still block on the wait result before writing response headers.

To consolidate the two functions I'd have to add more version checks throughout it. Would you prefer that?

**Comment by jlhawn** [comment]
@cpuguy83 I've added a commit which adds `WaitCondition` constants to the `api/types/container` package. Please take a look.

**Comment by jlhawn** [comment]
These build failures seem to be either Jenkins issues or flaky tests, can someone please trigger rebuilds for `experimental`, `janky`, and `windowsRS1`? 🙏

**Comment by jlhawn** [comment]
Rebased after conflicts with `docs/api/version-history.md`.

File: client/container_wait.go
Comment (tonistiigi) [inline client/container_wait.go]: Why does this return channels? I don't see anything in here returning multiple items. As soon as anything is written to channels it closes down.

File: container/state.go
Comment (tonistiigi) [inline container/state.go]: What needs this return channel? If we can avoid it there is no need for to rely on context canceling.

Comment (jlhawn) [inline container/state.go]: So if it didn't return a channel then it would have to block? Wouldn't you still want to pass in a context to cancel that?

The reason this returns a channel (where `WaitStop` did not before) is because this call is usually wrapped by a call to `something.ContainerWait(ctx, containerID)` which returns the pair `(chan Status, error)`. If there is an error before the wait can begin then the error value will be non-nil. Once the wait begins though, the call no longer blocks and you get a channel back and the caller *knows that there is a backend goroutine blocking on the container to exit or be removed*. This allows the caller do do other things, synchronously, before pulling the value out of the channel, such as:

- create a container
- get channel to wait for that container's next exit
- start the container
- block on receiving container's exit status

If this did not return a channel than there would be race conditions in this use case:

- create container
- spawn goroutine to wait for container's next exit
- start container
- block on waiting for container's exit

In this scenario it's possible that you start the container and it exits before that previously spawned goroutine actually gets a chance to wait for the container and would then wait for the exit which would never happen again.

In the case of the API handler for `ContainerWait`, the returned `(chan Status, error)` values allow us to always return response headers immediately once the wait begins (or return `No Such Container` in the case of an error). That allows for the *"okay, the backend is now waiting, I'll send you the response once I have it"* flow to work through the API.

File: container/state.go
Comment (jlhawn) [inline container/state.go]: It doesn't need a buffer in this case since blocking on it doesn't prevent this function from returning. It's buffered in the other case (when the condition and container is `not-running`) because we can avoid needing to spawn a goroutine to put it in the channel.

File: client/container_wait.go
Comment (jlhawn) [inline client/container_wait.go]: I guess this doesn't need to return an errors channel. The only reason why it does now is because there may be errors when decoding the response body. I could just log the error instead and set the response code to some non-zero value.

**Comment by jlhawn** [comment]
I've updated this PR again:
- Rebased to resolve yet another conflict with `docs/api/version-history.md`
- Added a new commit to address review comments from @tonistiigi

**Comment by cpuguy83** [comment]
```
15:49:24 The swagger spec at "api/swagger.yaml" is valid against swagger specification 2.0
15:49:26 The result of hack/generate-swagger-api.sh differs
15:49:26 
15:49:26  M api/types/container/container_wait.go
15:49:26 
15:49:26 Please update api/swagger.yaml with any api changes, then 
15:49:26 run `hack/generate-swagger-api.sh`.
```

**Comment by cpuguy83** [comment]
Ping @dnephin Do you know how we can add these new API objects to swagger so swagger-gen doesn't complain? Should we just move to a different file?

**Comment by dnephin** [comment]
`ContainerWait` is a generated struct (it's in `hack/generate-swagger-api.sh`), so you should never edit `api/types/container/container_wait.go` directly. Edit swagger to match what you want and generate it using  `hack/generate-swagger-api.sh`.

Is that what you mean?

**Comment by cpuguy83** [comment]
I mean it makes sense, but he's added some enums which need to be defined somewhere, and preferably in one place, but not sure if there is any facility for that in the swagger spec/generator.

**Comment by jlhawn** [comment]
> Edit swagger to match what you want ...

How would I do that? Edit the `api/swagger.yaml` file?

File: container/state.go
Comment (tonistiigi) [inline container/state.go]: > So if it didn't return a channel then it would have to block? Wouldn't you still want to pass in a context to cancel that?

Canceling a context if a caller doesn't need a result anymore is what the `context.WithCancel` is meant for. Requiring in the implementation that a context passed into the function must be canceled in order to not leak resources is a different constraint.

This `Wait` has no way to return early and provide an early error. If it is more valuable than confusing that you can have an error with `200` status code that doesn't need to leak into lower-level functions. If I look at the client side code(and builder)
https://github.com/moby/moby/pull/32237/files#diff-316ca96dd5834654f8dc1537e5377bc3R28 then I don't see this property of HTTP api used. I don't any callers checking the status of the returned channel and error channel differently. So do we actually handle that race?

File: container/state.go
Comment (tonistiigi) [inline container/state.go]: Can't it still leak a goroutine?

File: container/state.go
Comment (tonistiigi) [inline container/state.go]: Anyway, if it is needed for avoiding a race we can keep it, just document all the functions that require context to be canceled.

**Comment by dnephin** [comment]
Yes, sorry I noticed afterward it was just an enum. Putting them into a different file should work.

File: container/state.go
Comment (jlhawn) [inline container/state.go]: I guess another part of the contract with the caller would be that the caller *must* receive from the returned channel to avoid leaking any resources, since in this case the spawned goroutine does not exit until it is able to send on the channel. One could argue that the caller should always receive from the channel in which case it wouldn't matter whether the channel is buffered or not.

**Comment by jlhawn** [comment]
So, all of the `cli/command` files have disappeared. Do I have to open another PR now, against https://github.com/docker/cli ? I'm guessing that the api `client` package may soon disappear, too?

**Comment by jlhawn** [comment]
I'm going to preserve my `cli/command` changes in this diff file and deal with that later. I've rebased and this PR is now strictly an `api` and backend change.


[moby_cli.diff.txt](https://github.com/moby/moby/files/980618/moby_cli.diff.txt)

**Review by cpuguy83 — APPROVED** [review]
LGTM if green.

**Comment by jlhawn** [comment]
Tests *were* green, but I had to rebase to resolve still yet another conflict with `docs/api/version-history.md`.

**Comment by jlhawn** [comment]
screen shot 2017-05-11 at 10 51 02 am

💢💢💢💢💢💢💢💢💢💢💢💢

I'll keep rebasing to resolve the conflict, but I want to point out for posterity:

screen shot 2017-05-11 at 10 51 42 am

**Comment by WeiZhang555** [comment]
The code implementation looks good to me. 

I have only one doubt about the `next-exit`, as I can see there's no function using the `next-exit` feature currently, then should we keep it in the codes ? Though it costs only little redundant codes. :cat2:

**Comment by jlhawn** [comment]
> there's no function using the next-exit feature currently

There was, until the `cli` package was moved to https://github.com/docker/cli. See [this diff](https://github.com/moby/moby/files/980618/moby_cli.diff.txt) for the CLI changes for which I'll have to open another PR against that repository.

**Comment by WeiZhang555** [comment]
@jlhawn I see, LGTM :smile:

**Comment by thaJeztah** [comment]
ping @tonistiigi @dnephin PTAL!

**Comment by tonistiigi** [comment]
LGTM

File: client/container_wait.go
Comment (dnephin) [inline client/container_wait.go]: I have the same question.

Why does it return a `container.ContainerWaitOKBody` channel instead of just blocking?

File: api/server/router/container/container_routes.go
Comment (dnephin) [inline api/server/router/container/container_routes.go]: Instead of duplicating the blocking on channel and returning lines I think you could put this block in an `if !legacyBehavior {...}`

**Comment by jlhawn** [comment]
I'm currently rebasing and addressing @dnephin's comments. I want to expand on one of them since it has been asked already multiple times:

> Why does [the ContainerWait client method] return a container.ContainerWaitOKBody channel instead of just blocking?

This is done because it seemed like to quickest way to meet the new requirements we need for the Container Wait API which is to enable this flow:

1. Create a container in the case of `docker run ...` or have the container already exist in an exited state in the case of `docker start -a ...`.
2. Begin waiting for the container's *next* exit code (i.e., after it starts) or for it to be removed (in the case of `docker run --rm ...`.
3. Get *acknowledgement* from the daemon that it has begun waiting for that condition.
4. Start the container.
5. Continue to block on that wait condition.
6. Some time later, the container exits.
7. Receive the exit code.

- Point `2` is the reason why there is a `next-exit` wait condition.
- Point `3` is the reason why the API handler now returns response headers immediately.
- Being able to sync point `4` is the reason why the Client method returns a channel for the wait result - to be able to propagate the ack from point `3` and start the container without a race condition: If we didn't have a way of ensuring the order of `3` and `4` then it would be possible that the container is started and exits *before* the wait actually begins on the backend in which case `5` would block indefinitely.

I'm willing to change it to whatever we can all agree on. An alternative I have thought of is to go back to returning `(exitCode int, err error)` but to get the acknowledgement I mentioned we would need to have another argument to the method which is a channel which is maybe closed when the headers come back. That seems like an even more confusing interface to me though. Please make other suggestions if you have any ideas.

**Comment by jlhawn** [comment]
I have no idea why the `windowsRS1` build has failed:
```
can't load package: package github.com/docker/cli/cmd/docker: cannot find package "github.com/docker/cli/cmd/docker" in any of:
 	C:\go\src\github.com\docker\cli\cmd\docker (from $GOROOT)
 	C:\Users\ContainerAdministrator\AppData\Local\Temp\e5237f98-d5ea-4be8-bfb3-c68f9d2c2d38\src\github.com\docker\cli\cmd\docker (from $GOPATH)
```

**Review by dnephin — APPROVED** [review]
LGTM

A bit light on tests, but I guess we can address that later.

Are all 12 commits still relevant? Do you think some of them could be squashed now that this is ready to merge?

**Comment by jlhawn** [comment]
Rebased and squashed, it's now just 2 commits.

**Comment by thaJeztah** [comment]
Thanks!

**Comment by thaJeztah** [comment]
All 💚

**Comment by jlhawn** [comment]
The CLI changes are now in a PR against the `docker/cli` repo here: https://github.com/docker/cli/pull/101

File: client/container_wait.go
Comment (tiborvass) [reply client/container_wait.go]: Yeah why?

Comment (dnephin) [reply client/container_wait.go]: Response was here: https://github.com/moby/moby/pull/32237#issuecomment-301555524

**Pull request merged**
